### ISSUE TITLE
Support predictions in prior mode

### ISSUE THREAD

**Issue by jacobrgardner** [cornellius-gp/gpytorch#173]
Right now, it's essentially impossible to use an `ExactGP` to get predictions from the prior, because (1) we require training data in exact GP models, (2) in `train` mode we ensure that the supplied data is the same as the training data, and (3) in `eval` mode we compute the predictive posterior.

Some applications (like BayesOpt) don't necessarily start with training data at all, so getting predictions from the prior can be useful for choosing initial points, for example.

**Comment by Balandat** [comment]
This is a good point. By prior here I assume you mean a prior of the to-be-modeled function over the domain? Would this require a new kind of object?

In our applications, we often see that people would like to do "safe exploration", for which this would be quite useful (either using a prior based on some domain knowledge, or based on the posterior of a model from a related problem)

**Comment by jacobrgardner** [comment]
Yeah, I mean the prior over the to-be-modeled function. I've also had applications where it makes more sense to select points based on prior predictions to initialize BayesOpt than to choose randomly.

I don't think this will require a new object. **Technically** in code you can get prior predictions right now by creating a model using dummy tensors in `__init__` and then calling `forward` directly, which bypasses the check that we are calling on the training data in `model.train()` mode:

```python
likelihood = GaussianLikelihood(log_noise_bounds=(-5, 5))
model = ExactGPModel(torch.Tensor(0), torch.Tensor(0), likelihood)
prior_pred = model.forward(test_x)
prior_mean = prior_pred.mean()
prior_var = prior_pred.var()
prior_covar = prior_pred.covar().evaluate()
```
This interface is just obviously garbage.

**Comment by Balandat** [comment]
> I've also had applications where it makes more sense to select points based on prior predictions to initialize BayesOpt than to choose randomly.

See also my comment on the design doc about a heuristic we use to select starting points for gradient-based optimization.

@bkarrer, @bletham

**Comment by gpleiss** [comment]
@jacobrgardner @Balandat maybe it would make sense for ExactGPModel to accept `None` for the training data arguments? This would be prior mode.

**Comment by Balandat** [comment]
@jacobrgardner I also think that always doing this check by default is a little much - maybe we should have some kind of contextmanager that we can use to turn if off if we don't need the check to be done?

**Comment by gpleiss** [comment]
I would suggest:

- `model.eval()` with `None` for training data - prior mode
- `model.eval()` with training data - posterior mode
- `model.train()` - training mode, requires training data

@Balandat I like the idea of using a context manager for doing the checks. We also make the check in `eval` mode if the evaluation data matches the training data, and I think this check would also be nice to turn off as well.
